rfcs/2021-08-06-8619-framing-and-codecs-sources.md

```diff
@@ -0,0 +1,212 @@
+transform steps / input indirections, when basic transforms were used that are
+only concerned with encoding formats.
+
+## Prior Art
+
+[Tokio Codecs](https://docs.rs/tokio-util/0.6.7/tokio_util/codec/index.html)
+provide traits to conveniently convert from `AsyncRead`/`AsyncWrite` to
+`Stream`/`Sink`. These are currently used in custom implementations of sources
+to frame byte streams. We want to extend this existing facility to produce
+events from bytes directly.
+
+[Logstash Codec Plugins](https://www.elastic.co/guide/en/logstash/current/codec-plugins.html)
+are interesting since they operate on a higher level than what has been proposed
+in this RFC. They don't distinguish between a framing and codec stage, but
+rather have codecs that support framing (e.g. `line` codec), compression (e.g.
+`gzip_lines` codec) and encoding (e.g. `protobuf` codec). Supporting these kind
+of codecs could be an interesting future thought but would require an
+architectural change, especially to the internal representation of an `Event`.
+
+## Drawbacks
+
+It is possible that the proposed abstraction is too rigid, in a sense that it is
+not possible to cleanly separate these stages into "framing" and "codec"
+responsibilities. This limits what codecs can do, e.g. applying framing after
+decoding/decompression would not be possible as there are no means to "go back"
+to the framing stage. However, the proposed solution is still strictly better
+than the status quo, as it provides a consistent interface.
+
+# Alternatives Considered
+
+In a previous iteration of this RFC, codecs were implemented on the topology
+level and it was possible to chain multiple codecs to cover a wide range of
+scenarios, e.g. going from TCP stream -> decompress using `gzip` -> split by
+newline -> byte frame -> `json`.
+
+The advantage of that approach would be an easy mental model for the user since
+codecs would work for every source, and the flexibility of composing codecs.
+However, each codec implementation would need to accept any value (bytes,
+string, object, ...) and the internal event would need to represent a state for
+raw bytes / an unfinished event, meaning that the complexity to represent these
+combinations would propagate throughout the system.
+
+Since that approach would also allow users to configure senseless combinations
+of codecs, we decided to restrict this feature to going from bytes to `Event`
+through one framing and parsing method, which covers most use cases for which
+the framing/codec feature was requested.
+
+## Future Work
+
+In this first release, we plan to implement a variety of framing and codec
+options that cover the most common use cases. However, users might want to use a
+custom codec that only applies to their specific use case or is not supported by
+Vector yet. Adding a system that would allow custom-defined codecs could be
+considered in the future, for now this is can be accomplished by the `wasm` or
+`lua` transform.
```

Comment:
It might be worth calling out the idea of a `remap` codec, which seems like it would fit perfectly well into this model.

For example:

```toml
[decoding]
type = "remap"
src = """
. = parse_json!(.)
.nested = parse_json!(.nested)
.encoded = parse_base64!(.encoded)
"""
```

I wouldn't really recommend this as a first option for users, since it's more complex, has performance concerns, etc, but it could be a convenient escape hatch short of splitting out a transform.

It'd be extra nice if we could do a deeper remap integration like ensuring `.` is a byte string, etc, but I'll defer to remap experts on how feasible that is.

Reply:
I really like that one! That would pretty much solve the concern of "not being able to arbitrarily chain / compose codecs" by providing a reasonably performant escape hook that is still convenient to configure

Reply:
Added in efb2cda730ad19a7f7b056cab0bbba686b13e685